src/wp-includes/html-api/class-wp-html-open-elements.php

```diff
@@ -0,0 +1,317 @@
+	 *
+	 * @return WP_HTML_Token|null Last node in the stack of active formatting elements, if one exists, otherwise null.
+	 */
+	public function current_node() {
+		$current_node = end( $this->stack );
+
+		return $current_node ? $current_node : null;
+	}
+
+	/**
+	 * Returns whether an element is in a specific scope.
+	 *
+	 * ## HTML Support
+	 *
+	 * This function skips checking for the termination list because there
+	 * are no supported elements which appear in the termination list.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @see https://html.spec.whatwg.org/#has-an-element-in-the-specific-scope
+	 *
+	 * @param string   $tag_name         Name of tag check.
+	 * @param string[] $termination_list List of elements that terminate the search.
+	 * @return bool Whether the element was found in a specific scope.
+	 */
+	public function has_element_in_specific_scope( $tag_name, $termination_list ) {
+		foreach ( $this->walk_up() as $node ) {
+			if ( $node->node_name === $tag_name ) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Returns whether a particular element is in scope.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @see https://html.spec.whatwg.org/#has-an-element-in-scope
+	 *
+	 * @param string $tag_name Name of tag to check.
+	 * @return bool Whether given element is in scope.
+	 */
+	public function has_element_in_scope( $tag_name ) {
+		return $this->has_element_in_specific_scope(
+			$tag_name,
+			array(
+				// No termination elements are currently supported.
+			)
+		);
+	}
+
+	/**
+	 * Returns whether a particular element is in BUTTON scope.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @see https://html.spec.whatwg.org/#has-an-element-in-button-scope
+	 *
+	 * @param string $tag_name Name of tag to check.
+	 * @return bool Whether given element is in scope.
+	 */
+	public function has_element_in_button_scope( $tag_name ) {
+		return $this->has_element_in_specific_scope(
+			$tag_name,
+			array(
+				// No termination elements are currently supported.
+			)
+		);
+	}
+
+	/**
+	 * Returns whether a P is in BUTTON scope.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @see https://html.spec.whatwg.org/#has-an-element-in-button-scope
+	 *
+	 * @return bool
+	 */
+	public function has_p_in_button_scope() {
+		return $this->has_p_in_button_scope;
+	}
+
+	/**
+	 * Pops a node off of the stack of open elements.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @see https://html.spec.whatwg.org/#stack-of-open-elements
+	 *
+	 * @return bool Whether a node was popped off of the stack.
+	 */
+	public function pop() {
+		$item = array_pop( $this->stack );
+
+		if ( null === $item ) {
+			return false;
+		}
+
+		/*
+		 * When adding support for new elements, expand this switch to trap
+		 * cases where the precalculated value needs to change.
+		 */
+		switch ( $item->node_name ) {
+			case 'P':
+				$this->has_p_in_button_scope = $this->has_element_in_button_scope( 'P' );
+				break;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Repeatedly pop nodes off of the stack of open elements
```

Comment:
```suggestion
	 * Pops nodes off of the stack of open elements
```

Reply:
I'm incorporating your suggestion, but I wonder about it. I feel like `Repeatedly pops` captures more of the nuance while `Pops` leaves things more ambiguous. How strongly do you feel about this?

Reply:
I was thinking about how to retain `repeatedly` somewhere in the summary, but then I saw `until` and figured it might make `repeatedly` redundant?

The main thing of the suggestion was to use third-person singular verbs at the start of the summary per [existing standards](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#language). After many line/file-based commits to Core to enforce this over the years, there's been recent commits to enforce this in a "sweep" approach, and we regularly (always, we hope 😅 ) catch and enforce this during reviews of new code.

The main thing of the summary is that it's clear, so if there's a way to incorporate `repeatedly` somewhere else in the sentence, that would be the best of both worlds.

Reply:
It's mostly mixing enforced rules and enforcing personal style that gets me confused and tired. Thanks for clarifying. I'm going to keep the `repeatedly` out but change other things in things in this line.

It's interesting that both of us missed the more glaring defect in the sentence while reviewing the style use (which is fixed now in aae77bcaa9).